contracts/StorageAccessible.sol

```diff
@@ -23,4 +27,41 @@ contract StorageAccessible {
         }
         return result;
     }
+
+    /**
+     * @dev Performs a delegetecall on a targetContract in the context of self.
+     * Internally reverts execution to avoid side effects (making it static). Catches revert and returns encoded result as bytes.
+     * @param targetContract Address of the contract containing the code to execute.
+     * @param calldataPayload Calldata that should be sent to the target contract (encoded method name and arguments).
+     */
+    function simulateDelegatecall(
+        address targetContract,
+        bytes memory calldataPayload
+    ) public returns (bytes memory) {
+        bytes memory innerCall = abi.encodeWithSelector(
+            SIMULATE_DELEGATECALL_INTERNAL_SELECTOR,
+            targetContract,
+            calldataPayload
+        );
+        (, bytes memory response) = address(this).call(innerCall);
+        return response;
```

Comment:
Aren't there slight semantic differences between reverting and just returning an encoded revert message? Should you check the response for the `Error(string)` selector and revert here (to propagate the revert) or even encode if the call was successful or not?

Reply:
I now added a test case for delegateCalling a view function that reverts internally. It does get treated as a top level revert since as I assume the returned bytes are still the same. This is the behavior you expected, right?

Are you concerned about some return flag that would be set differently?

Reply:
> This is the behavior you expected, right?

Kind of. So actually this behaviour is AFAIK node dependent. From what I remember from `ethcontract`, Ganache returns a JSON RPC error when an `eth_call` reverts but Geth doesn't. So I imagine Truffle is designed to work with both, and will see the `Error(string)` selector in the response bytes and just assume it is a `revert` (which is how Geth would behave anyway).

> Are you concerned about some return flag that would be set differently?

So, I'm a bit more concerned about solidity code that calls `simulateDelegatecall` since it might subvert expectations that a `revert` in the `targetContract` call gets caught and not propagated. One use case that immediately comes to mind is writing an external contract (called A) to integrate with some contract (called B) that inherits from `StorageAccessible` and needs to read some storage data from contract B for a transaction. The fact that `simulateDelegatecall` does not propagate reverts might cause bugs because the developers aren't expecting this behaviour.

You can also argue that this is the caller's responsibility, in that case I would maybe add a bit more documentation about this.

Reply:
@nlordell can you take a look if https://github.com/gnosis/util-contracts/pull/31/commits/8a8f2ae1d83c8bfa655f56afa4fdc820ba2177b2 adds the behavior you were expecting?

Reply:
This is exactly the behaviour I was expecting.

Reply:
Hmm, what happens if it revers without a reason?

Reply:
Sorry, I think I'm getting carried away :stuck_out_tongue: , I think the way it is now is :ok_hand:

Reply:
<a href="https://gitme.me/image?url=https%3A%2F%2Fmedia2.giphy.com%2Fmedia%2F55kiN4lkOPJrKg9hdv%2Fgiphy.gif&token=keep_pushing" data-gitmeme-token="keep_pushing"><img src="https://media2.giphy.com/media/55kiN4lkOPJrKg9hdv/giphy.gif" title="Created by gitme.me with /keep_pushing"/></a>

Checkout https://github.com/gnosis/util-contracts/pull/31/commits/58f4ad28c9c311abb58eedd1cd91f10d1e3d9d49